Nprocs

miretskiy opened the pull request:
### span: make span.Frontier thread safe
Make `span.Frontier` thread safe by default.

Release justification: low impact data structure change.
Release note: None

### changefeedccl: parallelize event consumption
Previously, KV events were consumed and processed
by changefeed aggregator using a single, synchronous Go routine.
This PR makes it possible to run up to `changefeed.consumer_max_workers`
consumers to consume events concurrently.
The default of 0 keeps existing single threaded implementation.

Release justification: performance improvement, disabled by default.
Release note (enterprise change): Changefeeds can process events
concurrently, with up to `changefeed.consumer_max_workers` workers.

Commit:
span: make span.Frontier thread safe

Make `span.Frontier` thread safe by default.

Release justification: low impact data structure change.
Release note: None

Commit:
changefeedccl: parallelize event consumption

Previously, KV events were consumed and processed
by changefeed aggregator using a single, synchronous Go routine.
This PR makes it possible to run up to `changefeed.consumer_max_workers`
consumers to consume events concurrently.
The default of 0 keeps existing single threaded implementation.

Release justification: performance improvement, disabled by default.
Release note (enterprise change): Changefeeds can process events
concurrently, with up to `changefeed.consumer_max_workers` workers.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86658)
<!-- Reviewable:end -->

samiskin commented on pkg/ccl/changefeedccl/event_processing.go:
```diff
@@ -242,3 +319,124 @@ func (c *kvEventToRowConsumer) ConsumeEvent(ctx context.Context, ev kvevent.Even
+			idleTimer.Read = true
+
+			c.mu.Lock()
+			shutDown := c.mu.numWorkers > 1
```
Is it that bad to leave extra workers idle?  My assumption would've been that having an extra goroutine idling around is cheap.

miretskiy replied:
sure; but ... why...

samiskin replied:
My question is "why do extra work / have extra constraints for dynamic number of workers if it isn't providing value", where the extra work is the code around ramping up / down workers, the extra constraints is being slightly slower to ramp up capacity, while the value is something I'm not clear on since the only cost saving I can think of is having an extra idle routine.

miretskiy replied:
> My question is "why do extra work / have extra constraints for dynamic number of workers if it isn't providing value", where the extra work is the code around ramping up / down workers, the extra constraints is being slightly slower to ramp up capacity, while the value is something I'm not clear on since the only cost saving I can think of is having an extra idle routine.

I suppose this can be discussed once review (proper) is started...
Note however: it's not just about go routines; each one creates encoder, fetcher, "evaluator" -- these are
large datastructures, sometimes having their own caches, etc.

Timers are also cheap, and so is shutdown if worker didn't see any work in 2 minutes.

shermanCRL commented:
What's our theory as to the problems this addresses?

Is it strictly scale (“more processing is better”)? My question would be, is this actually a bottleneck? As a pipeline, I assume it’s as fast as the slowest part, which I’d guess would be the sink. Or, perhaps not, i.e. <big client> needs all the throughput we can give.

Or do we believe this might help with “stuckness”?

miretskiy commented on pkg/ccl/changefeedccl/changefeedbase/settings.go:
```diff
@@ -223,3 +223,13 @@ var UseMuxRangeFeed = settings.RegisterBoolSetting(
+	settings.TenantWritable,
+	"changefeed.consumer_max_workers",
+	"the maximum number of workers to use when processing events; 0 disables",
+	0,
```
util.ConstantWithMetamorphicTestRange("changefeed.consumer_max_workers", 0, 0, 32)

jayshrivastava committed:
changefeedccl: update parallel consumer to return async error

Previously, errors would not be returned by the parallel consumer until a
subsequent call to `feed.Next()`. This change makes it so that the parallel
consumer returns an error asynchronously via an error channel and cancels
its aggregator processor. This marks the underlying job as failed
and raises an error when expected.

Release justification: This change fixes a severe bug.
Release note: None

jayshrivastava committed:
do not merge: force parallel event consumer for testing

This change forces 32 parallel event consumer to be used
by changefeeds for testing purposes. Once CI passes, showing
that existing functionality is maintained by parallel consumers,
this commit can be dropped, reverting the number of consumers
to be a metamorphic constant.

Release justification: This will not be merged.
Release note: None
